test/core/iomgr/mpmcqueue_test.cc

```diff
@@ -0,0 +1,217 @@
+};
+
+class ConsumerThread {
+ public:
+  ConsumerThread(grpc_core::InfLenFIFOQueue* queue) : queue_(queue) {
+    thd_ = grpc_core::Thread(
+        "mpmcq_test_get_thd",
+        [](void* th) { static_cast<ConsumerThread*>(th)->Run(); }, this);
+  }
+  ~ConsumerThread() {}
+
+  void Start() { thd_.Start(); }
+  void Join() { thd_.Join(); }
+
+ private:
+  void Run() {
+    // count number of Get() called in this thread
+    int count = 0;
+
+    WorkItem* item;
+    while ((item = static_cast<WorkItem*>(queue_->Get())) != nullptr) {
+      count++;
+      GPR_ASSERT(!item->done);
+      item->done = true;
+    }
+
+    gpr_log(GPR_DEBUG, "ConsumerThread: %d times of Get() called.", count);
+  }
+  grpc_core::InfLenFIFOQueue* queue_;
+  grpc_core::Thread thd_;
+};
+
+static void test_get_empty(void) {
+  gpr_log(GPR_INFO, "test_get_empty");
+  grpc_core::InfLenFIFOQueue queue;
+  GPR_ASSERT(queue.count() == 0);
+  const int num_threads = 10;
+  ConsumerThread** consumer_thds = static_cast<ConsumerThread**>(
+      gpr_zalloc(num_threads * sizeof(ConsumerThread*)));
+
+  // Fork threads. Threads should block at the beginning since queue is empty.
+  for (int i = 0; i < num_threads; ++i) {
+    consumer_thds[i] = grpc_core::New<ConsumerThread>(&queue);
+    consumer_thds[i]->Start();
+  }
+
+  WorkItem** items = static_cast<WorkItem**>(
+      gpr_zalloc(THREAD_LARGE_ITERATION * sizeof(WorkItem*)));
+  for (int i = 0; i < THREAD_LARGE_ITERATION; ++i) {
+    items[i] = grpc_core::New<WorkItem>(i);
+    queue.Put(static_cast<void*>(items[i]));
+  }
+
+  gpr_log(GPR_DEBUG, "Terminating threads...");
+  for (int i = 0; i < num_threads; ++i) {
+    queue.Put(nullptr);
+  }
+  for (int i = 0; i < num_threads; ++i) {
+    consumer_thds[i]->Join();
+  }
+  gpr_log(GPR_DEBUG, "Checking and Cleaning Up...");
+  for (int i = 0; i < THREAD_LARGE_ITERATION; ++i) {
+    GPR_ASSERT(items[i]->done);
+    grpc_core::Delete(items[i]);
+  }
+  gpr_free(items);
+  for (int i = 0; i < num_threads; ++i) {
+    grpc_core::Delete(consumer_thds[i]);
+  }
+  gpr_free(consumer_thds);
+  gpr_log(GPR_DEBUG, "Done.");
+}
+
+static void test_FIFO(void) {
+  gpr_log(GPR_INFO, "test_FIFO");
+  grpc_core::InfLenFIFOQueue large_queue;
+  for (int i = 0; i < THREAD_LARGE_ITERATION; ++i) {
+    large_queue.Put(static_cast<void*>(grpc_core::New<WorkItem>(i)));
+  }
+  GPR_ASSERT(large_queue.count() == THREAD_LARGE_ITERATION);
+  for (int i = 0; i < THREAD_LARGE_ITERATION; ++i) {
+    WorkItem* item = static_cast<WorkItem*>(large_queue.Get());
+    GPR_ASSERT(i == item->index);
+    grpc_core::Delete(item);
+  }
+}
+
+static void test_many_thread(void) {
+  gpr_log(GPR_INFO, "test_many_thread");
+  const int num_work_thd = 10;
+  const int num_get_thd = 20;
+  grpc_core::InfLenFIFOQueue queue;
+  ProducerThread** work_thds = static_cast<ProducerThread**>(
+      gpr_zalloc(num_work_thd * sizeof(ProducerThread*)));
+  ConsumerThread** consumer_thds = static_cast<ConsumerThread**>(
```

Comment:
get_thds?

Let us keep consistent, either always using consumer/producer or put/get in this file.